src/klr/klr.cpp

```diff
@@ -7,7 +7,143 @@ void GetModuleDirectory(HMODULE module, LPWSTR szPath)
+{
+    int nIndex = lstrlen(pszStr) - 1;
+    do
+    {
+        if (pszStr[nIndex] == c)
+        {
+            break;
+        }
+    } while (--nIndex >= 0);
+
+    return nIndex;
+}
+
+bool StringsEqual(LPCWSTR const pszStrA, LPCWSTR const pszStrB)
+{
+    return ::_wcsicmp(pszStrA, pszStrB) == 0;
+}
+
+bool StringEndsWith(LPCWSTR const pszStr, LPCWSTR const pszSuffix)
+{
+    int nStrLen = lstrlen(pszStr);
+    int nSuffixLen = lstrlen(pszSuffix);
+    int nOffset = nStrLen - nSuffixLen;
+
+    if (nOffset < 0)
+    {
+        return false;
+    }
+
+    return StringsEqual(pszStr + nOffset, pszSuffix);
+}
+
+int LastPathSeparatorIndex(LPCWSTR const pszPath)
+{
+    int nLastSlashIndex = LastIndexOfChar(pszPath, L'/');
+    int nLastBackSlashIndex = LastIndexOfChar(pszPath, L'\\');
+    return max(nLastSlashIndex, nLastBackSlashIndex);
+}
+
+void GetParentDir(LPCWSTR const pszPath, LPWSTR const pszParentDir)
+{
+    int nLastSeparatorIndex = LastPathSeparatorIndex(pszPath);
+    if (nLastSeparatorIndex < 0)
+    {
+        StringCchCopyW(pszParentDir, MAX_PATH, L".");
+        return;
+    }
+
+    CopyMemory(pszParentDir, pszPath, (nLastSeparatorIndex + 1) * sizeof(WCHAR));
+    pszParentDir[nLastSeparatorIndex + 1] = L'\0';
+}
+
+void GetFileName(LPCWSTR const pszPath, LPWSTR const pszFileName)
+{
+    int nLastSeparatorIndex = LastPathSeparatorIndex(pszPath);
+
+    if (nLastSeparatorIndex < 0)
+    {
+        StringCchCopyW(pszFileName, MAX_PATH, pszPath);
+        return;
+    }
+
+    StringCchCopyW(pszFileName, MAX_PATH, pszPath + nLastSeparatorIndex + 1);
+}
+
+bool ExpandCommandLineArguments(int nArgc, LPWSTR* ppszArgv, int& nExpandedArgc, LPWSTR*& ppszExpandedArgv)
+{
+    if (nArgc == 0 || StringsEqual(ppszArgv[0], L"--appbase"))
+    {
+        return false;
+    }
+
+    nExpandedArgc = nArgc + 2;
+    ppszExpandedArgv = new LPWSTR[nExpandedArgc];
+    WCHAR szParentDir[MAX_PATH];
+
+    // "klr /path/App.dll arg1" --> "klr --appbase /path/ /path/App.dll arg1"
+    // "klr /path/App.exe arg1" --> "klr --appbase /path/ /path/App.exe arg1"
+    if (StringEndsWith(ppszArgv[0], L".exe") || StringEndsWith(ppszArgv[0], L".dll"))
+    {
+        GetParentDir(ppszArgv[0], szParentDir);
+
+        for (int i = 0; i < nExpandedArgc; ++i)
+        {
+            ppszExpandedArgv[i] = new WCHAR[MAX_PATH];
+            if (i >= 2)
+            {
+                StringCchCopyW(ppszExpandedArgv[i], MAX_PATH, ppszArgv[i - 2]);
+            }
+        }
+
+        StringCchCopyW(ppszExpandedArgv[0], MAX_PATH, L"--appbase");
+        StringCchCopyW(ppszExpandedArgv[1], MAX_PATH, szParentDir);
+
+        return true;
+    }
+
+    // "klr /path/project.json run" --> "klr --appbase /path/ Microsoft.Framework.ApplicationHost run"
+    // "klr /path/ run" --> "klr --appbase /path/ Microsoft.Framework.ApplicationHost run"
+    WCHAR szFileName[MAX_PATH];
+    GetFileName(ppszArgv[0], szFileName);
+    if (StringsEqual(szFileName, L"project.json"))
```

Comment:
Sorry I didn't notice the dot. Yeah, it reveals a bug and I just fixed it in the latest commit. After fix:
![capture](https://cloud.githubusercontent.com/assets/1383883/5750078/4f9b0584-9c04-11e4-91a1-faade2714d87.PNG)